File: scratch/BinaryTree/BinaryTree.java
import java.util.*;

public class BinaryTree {

    private static Node root;
    int size;

    static class Node {
        int data;
        Node left;
        Node right;

        public Node(int data) {
            this.data = data;
            left = null;
            right = null;
        }
    }

    public  BinaryTree() {
        size = 0;
        root = null;
    }

    public static boolean empty() {
        return root == null;
    }

    private static void inorderTraversal(Node root) {
        if (root != null) {
            inorderTraversal(root.left);
            System.out.print(root.data + " ");
            inorderTraversal(root.right);
        }
    }

    private static void postorderTraversal(Node root) {
        if (root != null) {
            postorderTraversal(root.left);
            postorderTraversal(root.right);
            System.out.println(root.data + " ");
        }
    }

    public static void insert(int data) {
        Node temp = root;
        Queue<Node> q = new LinkedList<Node>();
        q.add(root);

        while (!q.isEmpty()) {
            temp = q.peek();
            q.remove();

            if (temp.left == null) {
                temp.left = new Node(data);
                break;
            } else {
                q.add(temp.left);
            }

            if (temp.right == null) {
                temp.right = new Node(data);
                break;
            } else {
                q.add(temp.right);
            }
        }
    }


    public static void main(String[] args) {
        root = new Node(10);
        root.left = new Node(12);
        root.left.left = new Node(8);
        root.right = new Node(13);
        root.right.right = new Node(22);
        root.right.left = new Node(16);

        System.out.println("Before insertion");
        inorderTraversal(root);

        insert(18);
        System.out.println("\nAfter insertion");
        inorderTraversal(root);
    }
}

File: scratch/BinaryTree/binary_tree.py
class Node(object):
    def __init__(self, data=None, left_node=None, right_node=None):
        self.data = data
        self.left_node = left_node
        self.right_node = right_node


class BinaryTree(object):
    def __init__(self, root=None):
        self.root = root

    @property
    def is_empty(self) -> bool:
        return self.root is None

    @property
    def get_root(self) -> Node:
        return self.root

    def inorder(self, node):
        if node is not None:
            self.inorder(node.left_node)
            print(node.data)
            self.inorder(node.right_node)

    def preorder(self, node):
        if node is not None:
            print(node.data)
            self.preorder(node.left_node)
            self.preorder(node.right_node)

    def postorder(self, node):
        if node is not None:
            self.postorder(node.left_node)
            self.postorder(node.right_node)
            print(node.data)

    def breadth_first_search(self, value: any) -> bool:
        if self.root is None:
            return False
        queue = [self.root]

        while queue:
            node = queue.pop()
            if node.left_node != value:
                queue.append(node.left_node)
            else:
                print('Value found!')
                return True

            if node.right_node != value:
                queue.append(node.right_value)
            else:
                print('Value found!')
                return True

        print('Value could not be found')
        return False

    def breadth_first_traversal(self):
        queue = [self.root]
        visited = []

        while queue:
            node = queue.pop()
            if node.left_node not in visited:
                visited.append(node.left_node)
                queue.append(node.left_node)
                continue

            if node.right_node not in visited:
                visited.append(node.right_node)
                queue.append(node.right_node)
                continue

        # Print the nodes
        for node in visited:
            print(node.data)

    def insert(self, new_data: any):
        queue = [self.root]

        while queue:
            node = queue.pop()
            if node.left_node is None:
                node.left_node = Node(new_data, None, None)
                return
            else:
                queue.append(node.left_node)

            if node.right_node is None:
                node.right_node = Node(new_data, None, None)
                return
            else:
                queue.append(node.right_node)


def main():
    root_node = Node(1, None, None)
    bt = BinaryTree(root_node)
    bt.insert(4)
    bt.insert(5)
    bt.insert(11)
    bt.inorder(bt.root)


main()
